### PR TITLE
Add RouteTableResourceGroup for Azure cloud provider

### PR DESCRIPTION
**Pull request by titilambert**
**What type of PR is this?**
> Uncomment only one, leave it on its own line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test

/kind feature
> /kind flake

**What this PR does / why we need it**:

This PR adds a new parameter in the Azure Cloud Provider config file.
The new parameter is RouteTableResourceGroup. It has the same behavior as the VNetResourceGroup has for Vnet. Because sometimes, you want to use an Azure RouteTable which is not is in the same Resource Group as your main Resource Group

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Azure Cloud Provider: Add new parameter `RouteTableResourceGroup`
```

### CODE REVIEW

**Comment by titilambert** [comment]
Could this PR be merged also in 1.11.x ?

File: pkg/cloudprovider/providers/azure/azure_routes.go [hunk 1]
```diff
@@ -97,6 +97,12 @@ func (az *Cloud) createRouteTableIfNotExists(clusterName string, kubeRoute *clou
 }
 
 func (az *Cloud) createRouteTable() error {
+	var rg string
+	if len(az.RouteTableResourceGroup) > 0 {
```
Comment (yastij) [inline pkg/cloudprovider/providers/azure/azure_routes.go]: Can you add a test case to cover defaulting the RG

File: pkg/cloudprovider/providers/azure/azure_backoff.go [hunk 2]
```diff
@@ -262,7 +262,14 @@ func (az *Cloud) CreateOrUpdateRouteTableWithRetry(routeTable network.RouteTable
 		ctx, cancel := getContextWithCancel()
 		defer cancel()
 
-		resp, err := az.RouteTablesClient.CreateOrUpdate(ctx, az.ResourceGroup, az.RouteTableName, routeTable)
+		var rg string
+		if len(az.RouteTableResourceGroup) > 0 {
```
Comment (yastij) [inline pkg/cloudprovider/providers/azure/azure_backoff.go]: The defaulting should be done outside, as this will avoid duplication

**Review by yastij — CHANGES_REQUESTED** [review]
Left some comments, @feiskyer have you heard of such a use case ?

File: pkg/cloudprovider/providers/azure/azure_backoff.go (hunk 2 above)
Comment (titilambert) [reply pkg/cloudprovider/providers/azure/azure_backoff.go]: Changed !

File: pkg/cloudprovider/providers/azure/azure_routes.go (hunk 1 above)
Comment (titilambert) [reply pkg/cloudprovider/providers/azure/azure_routes.go]: I'm pretty a newbee. Where can I find an example ? Thanks !

Comment (titilambert) [reply pkg/cloudprovider/providers/azure/azure_routes.go]: Done !

**Comment by feiskyer** [comment]
> Left some comments, @feiskyer have you heard of such a use case ?

Not yet. @brendandburns @khenidak @justaugustus have some of you heard of such use cases?

**Comment by titilambert** [comment]
I have this issue because the Vnet and the Route Table are in a different Resource Group from  Kubernetes one (the route table was created manually)

File: pkg/cloudprovider/providers/azure/azure_backoff.go [hunk 3]
```diff
@@ -285,7 +285,14 @@ func (az *Cloud) DeleteRouteWithRetry(routeName string) error {
 		ctx, cancel := getContextWithCancel()
 		defer cancel()
 
-		resp, err := az.RoutesClient.Delete(ctx, az.ResourceGroup, az.RouteTableName, routeName)
+		var rg string
+		if len(az.RouteTableResourceGroup) > 0 {
```
Comment (yastij) [inline pkg/cloudprovider/providers/azure/azure_backoff.go]: This should be also refactored

**Comment by suker200** [comment]
Hi, I met this case too. Do we have any progress on this ?

**Comment by justaugustus** [comment]
Alright, we've got multiple requests for this, so I'm happy to approve once it gets fixed up.
@titilambert -- can you rebase and address the comments?

**Comment by suker200** [comment]
hi, can I make another PR the same like this with rebase to push this progress ?

**Comment by yastij** [comment]
@suker200 - I agree with @justaugustus, after giving some thoughts it seems like there's reasonable use cases, feel free to open a PR

**Comment by feiskyer** [comment]
Close this one in favor of #75580

/close

**Pull request closed without merging**